### PR TITLE
Story/sc 1110

### PR DESCRIPTION
**Pull request by abnepc**
## Jira Story
[SC-1110](https://epcpower.atlassian.net/browse/SC-1110)
## About
Added scale factor column and enumerations worksheet to static modbus spreadsheet output
## Associated Pull Requests
*   [PR-8](https://github.com/epcpower/rm-dc/pull/8)

## Update Changelog

*   [x] Changelog updated

## Reproduction Steps
N/A
## Validation
Follow the steps for validation in the associated PR

[SC-1110]: https://epcpower.atlassian.net/browse/SC-1110?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .vscode/settings.json
Comment (gordon-epc) [inline .vscode/settings.json:1]: Remove this file. The correct way is to change the `.gitignore`, but for this repo it's a little more complicated. We'll leave the `.gitignore` modification for another PR.

File: src/epcpm/staticmodbustoxls.py [hunk 1]
```diff
@@ -105,26 +183,67 @@ class Root:
...
-
         Returns:
             workbook: generated Excel workbook
         """
         workbook = openpyxl.Workbook()
         workbook.remove(workbook.active)
-        worksheet = workbook.create_sheet()
-        worksheet.append(field_names.as_filtered_tuple(self.column_filter))
-
+        modbus_worksheet = workbook.create_sheet("Static Modbus Data")
```
Comment (gordon-epc) [inline src/epcpm/staticmodbustoxls.py:144]:
```suggestion
        modbus_worksheet = workbook.create_sheet("Static Modbus Map")
```

**Review by gordon-epc — CHANGES_REQUESTED** [review]

File: src/epcpm/staticmodbustoxls.py (hunk 1 above)
Comment (abnepc) [reply src/epcpm/staticmodbustoxls.py:144]: [827e071](https://github.com/epcpower/pm/pull/71/commits/827e0716a3fb22c7d2e4132dd00d59cb734c070a)

File: .vscode/settings.json
Comment (abnepc) [reply .vscode/settings.json:1]: [827e071](https://github.com/epcpower/pm/pull/71/commits/827e0716a3fb22c7d2e4132dd00d59cb734c070a)

File: src/epcpm/staticmodbustoxls.py [hunk 2]
```diff
@@ -42,11 +46,93 @@ class Fields(epcpm.pm_helper.FieldsInterface):
...
+        enumerators_list (list): list of enumerator dicts
+
+    Returns:
+        list: list of lists where each list contains the enumerator `name` and `value`
+    """
+    cleaned_enumerators_list = []
+    for enumerator in enumerators_list:
+        cleaned_enumerators_list.append([enumerator.name, enumerator.value])
+    return cleaned_enumerators_list
+
```
Comment (gordon-epc) [inline src/epcpm/staticmodbustoxls.py]: Don't need this anymore.
```suggestion
```

File: src/epcpm/staticmodbustoxls.py [hunk 3]
```diff
@@ -105,26 +183,67 @@ class Root:
...
+        enumeration_worksheet.append(["Name", "Value"])
+
+        scale_factor_from_uuid = build_uuid_scale_factor_dict(
+            points=self.wrapped.children,
+            parameter_uuid_finder=self.parameter_uuid_finder
+        )
+        enumerators_from_uuid = build_enumerators_dict(
+            points=self.wrapped.children,
+            parameter_uuid_finder=self.parameter_uuid_finder
+        )
```
Comment (gordon-epc) [inline src/epcpm/staticmodbustoxls.py]:
```suggestion
```

File: src/epcpm/staticmodbustoxls.py [hunk 4]
```diff
@@ -105,26 +183,67 @@ class Root:
...
             )
             rows = builder.gen()
             for row in rows:
-                worksheet.append(row.as_filtered_tuple(self.column_filter))
+                modbus_worksheet.append(row.as_filtered_tuple(self.column_filter))
+
+        for enumeration_name, enumerators in enumerators_from_uuid.items():
+            enumeration_worksheet.append([enumeration_name])
+            for enumerator in enumerators:
+                enumeration_worksheet.append(enumerator)
```
Comment (gordon-epc) [inline src/epcpm/staticmodbustoxls.py]:
```suggestion
        for enumeration in enumerations:
            for child in enumeration.children:
                enumeration_worksheet.append([enumeration.name, child.name, child.value])
```

File: src/epcpm/staticmodbustoxls.py [hunk 5]
```diff
@@ -105,26 +183,67 @@ class Root:
...
         """
         workbook = openpyxl.Workbook()
         workbook.remove(workbook.active)
-        worksheet = workbook.create_sheet()
-        worksheet.append(field_names.as_filtered_tuple(self.column_filter))
-
+        modbus_worksheet = workbook.create_sheet("Static Modbus Map")
+        enumeration_worksheet = workbook.create_sheet("Enumerations")
+        modbus_worksheet.append(field_names.as_filtered_tuple(self.column_filter))
+        enumeration_worksheet.append(["Name", "Value"])
```
Comment (gordon-epc) [inline src/epcpm/staticmodbustoxls.py]:
```suggestion
        enumeration_worksheet.append(["Enumerator", "Name", "Value"])
```

File: src/epcpm/staticmodbustoxls.py [hunk 6]
```diff
@@ -105,26 +183,67 @@ class Root:
...
+
+        scale_factor_from_uuid = build_uuid_scale_factor_dict(
+            points=self.wrapped.children,
+            parameter_uuid_finder=self.parameter_uuid_finder
+        )
+        enumerators_from_uuid = build_enumerators_dict(
+            points=self.wrapped.children,
+            parameter_uuid_finder=self.parameter_uuid_finder
+        )
+        enumerations = self.collect_enumerations()
```
Comment (gordon-epc) [inline src/epcpm/staticmodbustoxls.py:153]:
```suggestion
        enumerations = self.collect_enumerations()
        enumerators_from_uuid = {}
        for enumeration in enumerations:
            enumerators_from_uuid[enumeration.uuid] = enumeration

```

File: src/epcpm/staticmodbustoxls.py [hunk 7]
```diff
@@ -105,26 +183,67 @@ class Root:
...
         for member in self.wrapped.children:
             builder = builders.wrap(
                 wrapped=member,
                 parameter_uuid_finder=self.parameter_uuid_finder,
+                scale_factor_from_uuid=scale_factor_from_uuid,
+                enumerators_from_uuid=enumerations
```
Comment (gordon-epc) [inline src/epcpm/staticmodbustoxls.py]:
```suggestion
                enumerators_from_uuid=enumerators_from_uuid
```

File: src/epcpm/staticmodbustoxls.py [hunk 8]
```diff
@@ -163,6 +284,11 @@ def gen(self) -> typing.List[Fields]:
                 access_level = self.parameter_uuid_finder(parameter.access_level_uuid)
                 row.access_level = access_level.value
 
+            if self.wrapped.factor_uuid is not None:
+                row.scale_factor = self.parameter_uuid_finder(
+                    self.scale_factor_from_uuid[self.wrapped.factor_uuid].parameter_uuid
+                ).abbreviation
+
```
Comment (gordon-epc) [inline src/epcpm/staticmodbustoxls.py:275]:
```suggestion

            if self.wrapped.enumeration_uuid is not None:
                row.enumerator = self.parameter_uuid_finder(
                    self.enumerators_from_uuid[self.wrapped.enumeration_uuid].uuid
                ).name
```

**Review by gordon-epc — CHANGES_REQUESTED** [review]

File: src/epcpm/staticmodbusmodel.py [hunk 9]
```diff
@@ -90,7 +90,6 @@ def create_size_attribute(default=0):
         ),
     )
 
-
```
Comment (gordon-epc) [inline src/epcpm/staticmodbusmodel.py]: No other changes in this file, so undo this change.

**Review by gordon-epc — CHANGES_REQUESTED** [review]

File: src/epcpm/staticmodbusmodel.py (hunk 9 above)
Comment (abnepc) [reply src/epcpm/staticmodbusmodel.py]: [f1ae13b](https://github.com/epcpower/pm/pull/71/commits/f1ae13b27cd7b0f7ca2d484868d10e62f05021c6)

File: src/epcpm/staticmodbustoxls.py [hunk 10]
```diff
@@ -105,26 +134,84 @@ class Root:
...
+                if row.enumerator is not None:
+                    max_row = modbus_worksheet.max_row
+                    if row.enumerator in enumeration_cell_coordinates.keys():
+                        modbus_worksheet[MODBUS_ENUMERATOR_COLUMN +f"{max_row}"].hyperlink = (
+                            enumeration_cell_coordinates[row.enumerator]
+                        )
 
         return workbook
 
+    def collect_enumerations(self):
```
Comment (gordon-epc) [inline src/epcpm/staticmodbustoxls.py]: Can you add a return type hint and docstring for `collect_enumerations`, please?

**Review by gordon-epc — CHANGES_REQUESTED** [review]

File: src/epcpm/staticmodbustoxls.py (hunk 10 above)
Comment (abnepc) [reply src/epcpm/staticmodbustoxls.py]: [5aea6b6](https://github.com/epcpower/pm/pull/71/commits/5aea6b6671bfebe84a18b153f896d3b0bfbebdf3)

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -60,7 +60,6 @@ jobs:
       - name: python commands
         run: |
           set -vx
-          poetry run black --check --diff .
```
Comment (billiam-epc) [inline .github/workflows/ci.yml]: why was this taken out? shouldn't it be running the formatter?

File: src/epcpm/cli/main.py [hunk 12]
```diff
@@ -124,7 +124,7 @@ def docs(
     generate_formatted_output: bool,
 ) -> None:
     """
```
Comment (billiam-epc) [inline src/epcpm/cli/main.py]: I notice this  `"""` is on a different line than the others, is this going to fly for the other team?

File: src/epcpm/parameterstointerface.py [hunk 13]
```diff
@@ -1691,8 +1707,8 @@ def create_common_initializers(
             f".variable = {sunspec_variable},",
             f".getter = {sunspec_getter},",
             f".setter = {sunspec_setter},",
-            f".handGetter = {hand_coded_sunspec_getter_function},",
```
Comment (billiam-epc) [inline src/epcpm/parameterstointerface.py]: Commented out code.

File: src/epcpm/staticmodbustoxls.py [hunk 14]
```diff
@@ -42,11 +46,44 @@ class Fields(epcpm.pm_helper.FieldsInterface):
...
+
+    Args:
+        points (list): a list of parameter dicts
+        parameter_uuid_finder (function): function to identify a node given a UUID
+
+    Returns:
+        dict: a dict where the key is the parameter UUID and the value is the
+        parameter dict associated with the UUID
+    """
+    # TODO: CAMPid 45002738594281495565841631423784
```
Comment (billiam-epc) [inline src/epcpm/staticmodbustoxls.py:73]: please tell me this is a Kyle hold over

File: src/epcpm/staticmodbustoxls.py [hunk 15]
```diff
@@ -105,26 +134,91 @@ class Root:
...
+
+        Returns:
+            list: a list of enumerations
+        """
+        collected = []
+
+        if self.parameter_model is None:
+            return collected
+
+        def collect(node, collected):
```
Comment (billiam-epc) [inline src/epcpm/staticmodbustoxls.py]: can you provide type hints and return hint for this?

File: src/epcpm/cli/main.py (hunk 12 above)
Comment (gordon-epc) [reply src/epcpm/cli/main.py]: Only focus on `staticmodbustoxls.py` and `CHANGELOG.md`. The other diff files are due to the mix of merging `develop` and `feature/rm-dc branches`.

File: src/epcpm/parameterstointerface.py (hunk 13 above)
Comment (gordon-epc) [reply src/epcpm/parameterstointerface.py]: Only focus on `staticmodbustoxls.py` and `CHANGELOG.md`. The other diff files are due to the mix of merging `develop` and `feature/rm-dc branches`.

File: src/epcpm/staticmodbustoxls.py (hunk 14 above)
Comment (gordon-epc) [reply src/epcpm/staticmodbustoxls.py:73]: There are a lot of moving parts with this repo right now. I'll have to check if this is still necessary.

File: src/epcpm/staticmodbustoxls.py (hunk 15 above)
Comment (gordon-epc) [reply src/epcpm/staticmodbustoxls.py]: I know I said no docstring necessary, but Bill's right about the type hints.

Comment (abnepc) [reply src/epcpm/staticmodbustoxls.py]: [c5a4ceb](https://github.com/epcpower/pm/pull/71/commits/c5a4ceb117bf783d9393156b40acc31c6b51bdba)
